### PR TITLE
API Use concrete Handler implementations for Spam and Approve bulk editing

### LINKED ISSUE #231: Update to match colymba/GridFieldBulkEditingTools changes
Bringing in CWP 2.0.0-rc1, I cannot edit a page in the CMS. The error thrown is:

```
PHP Fatal error:  Bulk action handler not found:  in /sites/smalltest2/releases/20180223155356/vendor/colymba/gridfield-bulk-editing-tools/src/BulkManager/BulkManager.php on line 113
```

This is because the `addBulkAction()` method has a totally different implementation with the version pulled in with the recipe, that is SS4 compatible. Basically, [this whole section](https://github.com/silverstripe/silverstripe-comments/blob/master/src/Admin/CommentsGridFieldConfig.php) needs to be rewritten

To reproduce:

1) install CWP recipe 2.0.0-rc1
2) once you've hacked around with `composer.json` to pull in the up-to-date version of `colymba/gridfield-bulk-editing-tools`, visit `admin/pages` section
3) click on any page to edit it, you will get an Internal Server Error

### PR DESCRIPTION
**Pull request by robbieaverill**
Resolves #231

Doesn't require https://github.com/colymba/GridFieldBulkEditingTools/pull/163, but this PR helps the UI slightly

### CODE REVIEW

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace SilverStripe\Comments\Admin\CommentsGridFieldBulkAction;
+
+use Colymba\BulkManager\BulkAction\Handler;
+use SilverStripe\Core\Convert;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+
+/**
+ * A {@link GridFieldBulkActionHandler} for bulk marking comments as spam
```
Comment (andrewandante) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: Someone did the spam handler first, I think 😉

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    protected $label = 'Approve';
+
+    /**
+     * @param  HTTPRequest $request
+     * @return HTTPResponse
+     */
+    public function index(HTTPRequest $request)
+    {
+        $ids = array();
```
Comment (andrewandante) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: Any reason this isn't  `[]`?

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        foreach ($this->getRecords() as $record) {
+            array_push($ids, $record->ID);
+            $record->markApproved();
+        }
+
+        $response = new HTTPResponse(Convert::raw2json(array(
```
Comment (andrewandante) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: Same here re: `[]`

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            'done' => true,
+            'records' => $ids
```
Comment (andrewandante) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: Trailing comma?

File: src/Admin/CommentsGridFieldBulkAction/SpamHandler.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    protected $label = 'Spam';
+
+    /**
+     * @param  HTTPRequest $request
+     * @return HTTPResponse
+     */
+    public function index(HTTPRequest $request)
+    {
+        $ids = array();
```
Comment (andrewandante) [inline src/Admin/CommentsGridFieldBulkAction/SpamHandler.php]: Apply `[]` changes here too 😄

**Comment by robbieaverill** [comment]
Thanks @andrewandante, updated to fix up the points in your comments

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * @param  HTTPRequest $request
+     * @return HTTPResponse
+     */
+    public function index(HTTPRequest $request)
+    {
+        $ids = [];
+
+        foreach ($this->getRecords() as $record) {
+            array_push($ids, $record->ID);
```
Comment (dhensby) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: I don't have a problem explicitly with using the `array_push` method, but I normally use (and assume the SS code base does to) `$ids[] = $record->ID;` for pushing to an array

File: src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            $record->markApproved();
+        }
+
+        $response = new HTTPResponse(Convert::raw2json([
+            'done' => true,
+            'records' => $ids,
+        ]));
+
+        $response->addHeader('Content-Type', 'text/json');
```
Comment (dhensby) [inline src/Admin/CommentsGridFieldBulkAction/ApproveHandler.php]: It's `application/json`?

---

src: http://www.ietf.org/rfc/rfc4627.txt

**Comment by robbieaverill** [comment]
Yeah nice, thanks @dhensby!

**Pull request merged**